[net8.0] Merge latest main to net8

rmarinho opened the pull request:
### Description of Change

Brings latest changes on main to the net8.0 branch for next release.

jsuarezruiz committed:
Fix issue using FillRule with Paths (#14493)

* Fix issue using FillRule with Paths

* Fix build errors

* Fixed issue on iOS

* Fix wrong var name

* Updated code

* Small change

jsuarezruiz committed:
[Core] Fix Border margin issue (#14402) Fixes #7764

* Fox Border margin issue

* Auto-format source code

---------

Co-authored-by: GitHub Actions Autoformatter <autoformat@example.com>

mattleibow committed:
Added ICommandMapper for extension methods (#14545)

* Added ICommandMapper for extension methods

* No need for breaking changes

* Keep the previous method parameter

* Add a way to replace without base

* Remove When

* Update CommandMapperExtensionTests.cs

* Update PropertyMapperExtensionTests.cs

mattleibow committed:
Improve some test methods (#14753)

* Improve some test methods

* Add constraints

* object

* :)

PureWeen committed:
Call the correct Disconnect method from DisconnectHandler (#14757)

PureWeen committed:
Switch to AppWindowTitleBar From SetTitleBar (#14517)

* Convert TitleBar implemenation over to AppTitleBar

* Update WindowHandler.Windows.cs

* - add more samples and organize code

jfversluis committed:
Update All (most) BindableProperty XML docs (#14704)

github-actions[bot] committed:
[create-pull-request] automated change (#14776)

Co-authored-by: github-actions[bot] <41898282+github-actions[bot]@users.noreply.github.com>

dependabot[bot] committed:
Bump Mono.Cecil from 0.11.4 to 0.11.5 (#14768)

Bumps [Mono.Cecil](https://github.com/jbevain/cecil) from 0.11.4 to 0.11.5.
- [Release notes](https://github.com/jbevain/cecil/releases)
- [Commits](https://github.com/jbevain/cecil/compare/0.11.4...0.11.5)

---
updated-dependencies:
- dependency-name: Mono.Cecil
  dependency-type: direct:production
  update-type: version-update:semver-patch
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

PureWeen committed:
Set menuItem show action back to Never (#14623)

* Set menuItem show action back to Never

* Auto-format source code

* - fix iOS handler registration

---------

Co-authored-by: GitHub Actions Autoformatter <autoformat@example.com>

PureWeen committed:
Centralize logic for handling logical children (#14132)

jsuarezruiz committed:
Use Background as Fill, if Fill is null in Shapes (#14216) Fixes #13688

* Fix issues with shapes shadows on Android

* Updated sample

* Added comments

jonathanpeppers committed:
[core] use ToUpperInvariant() / ToLowerInvariant() (#14773)

Context: https://learn.microsoft.com/dotnet/fundamentals/code-analysis/quality-rules/ca1311
Context: https://learn.microsoft.com/previous-versions/dotnet/articles/ms994325(v=msdn.10)#the-turkish-example

In 8a58e01f, I found & fixed some performance issues around
culture-aware strings. This takes this further by making `CA1311` a
build error.

As mentioned in "the turkish example", doing something like:

    string text = something.ToUpper();
    switch (text) { ... }

Can actually cause unexpected behavior in Turkish locales, because in
Turkish, the character I (Unicode 0049) is considered the upper case
version of a different character ý (Unicode 0131), and i (Unicode
0069) is considered the lower case version of yet another character Ý
(Unicode 0130).

`ToLowerInvariant()` and `ToUpperInvariant()` are also better for
startup performance as it avoids loading the current culture.

I only found one place that actually *would* want to use the current
culture, which is in `CaseConverter.cs`.

PureWeen committed:
Create better helpers for attaching and fix Window (#14772)

* Create better helpers for attaching and fix Window

* - add a few more helpers into the basement

* - fix compile error on swipeview

* Auto-format source code

* - fix focus tests

* - move methods up

* - fix some left over code

---------

Co-authored-by: GitHub Actions Autoformatter <autoformat@example.com>

mattleibow committed:
Don't measure text spans if type is HTML (#14705)

mattleibow committed:
Support reading fonts from files (#14783)

mattleibow committed:
Correctly re-apply font formatting to HTML text (#14701)

Setting the text on iOS resets all formatting, so we need to re-apply size, color and other properties.

PureWeen committed:
Use new methods for checking colors (#14809)

mattleibow committed:
Transition away from implicit package references (#14770)

PureWeen committed:
Make sure to call `InitializeViewHandler` when creating a handler (#14814)

* Initialize View Handler after creating it

* - remove attach and run that uses tohandler

jsuarezruiz committed:
Avoid Android OnDestroy  lifecycle event firing twice (#14799)

rmarinho committed:
Merge branch 'main' into merge-main-net8

# Conflicts:
#	src/Core/src/PublicAPI/net-android/PublicAPI.Unshipped.txt

mattleibow committed:
Fix some sample XAML so the text wraps (#14821)

mattleibow committed:
Add some dummy images to the sandbox (#14815)

PureWeen committed:
Propagate Navigation Events through PageContainer (#14523)

* Propagate Navigation Events through PageContainer

* - clean up tests

* Auto-format source code

* - remove no longer applicable test

---------

Co-authored-by: GitHub Actions Autoformatter <autoformat@example.com>

jsuarezruiz committed:
[Windows] Update CollectionView changing ItemsLayout  (#14532)

* Added sample

* Testing

* Fix the issue

* Small refactoring

* Refactored code

---------

Co-authored-by: Rui Marinho <me@ruimarinho.net>

rmarinho committed:
Merge branch 'main' into merge-main-net8

rmarinho commented on src/Core/src/PublicAPI/net-android/PublicAPI.Shipped.txt:
```diff
@@ -2206,7 +2206,6 @@ override Microsoft.Maui.MauiAppCompatActivity.OnActivityResult(int requestCode,
 override Microsoft.Maui.MauiAppCompatActivity.OnBackPressed() -> void
 override Microsoft.Maui.MauiAppCompatActivity.OnConfigurationChanged(Android.Content.Res.Configuration! newConfig) -> void
 override Microsoft.Maui.MauiAppCompatActivity.OnCreate(Android.OS.Bundle? savedInstanceState) -> void
-override Microsoft.Maui.MauiAppCompatActivity.OnDestroy() -> void
```
@mattleibow shouldn't this be marked as remove on Unshipped ?

mattleibow replied:
It probably should have, but it is an override so we can let it slide...

rmarinho commented on src/Core/src/PublicAPI/net-android/PublicAPI.Unshipped.txt:
```diff
@@ -37,6 +44,13 @@ override Microsoft.Maui.SizeRequest.GetHashCode() -> int
+static Microsoft.Maui.CommandMapperExtensions.AppendToMapping<TVirtualView, TViewHandler>(this Microsoft.Maui.ICommandMapper<TVirtualView, TViewHandler>! commandMapper, string! key, System.Action<TViewHandler, TVirtualView, object?>! method) -> void
+static Microsoft.Maui.CommandMapperExtensions.AppendToMapping<TVirtualView, TViewHandler>(this Microsoft.Maui.ICommandMapper<Microsoft.Maui.IElement!, Microsoft.Maui.IElementHandler!>! commandMapper, string! key, System.Action<TViewHandler, TVirtualView, object?>! method) -> void
+static Microsoft.Maui.CommandMapperExtensions.ModifyMapping<TVirtualView, TViewHandler>(this Microsoft.Maui.ICommandMapper<TVirtualView, TViewHandler>! commandMapper, string! key, System.Action<TViewHandler, TVirtualView, object?, System.Action<Microsoft.Maui.IElementHandler!, Microsoft.Maui.IElement!, object?>?>! method) -> void
+static Microsoft.Maui.CommandMapperExtensions.ModifyMapping<TVirtualView, TViewHandler>(this Microsoft.Maui.ICommandMapper<Microsoft.Maui.IElement!, Microsoft.Maui.IElementHandler!>! commandMapper, string! key, System.Action<TViewHandler, TVirtualView, object?, System.Action<Microsoft.Maui.IElementHandler!, Microsoft.Maui.IElement!, object?>?>! method) -> void
```
@mattleibow but I m adding the api here too not sure why is complaining

rmarinho committed:
Adding missing api

Commit:
Auto-format source code

rmarinho commented:
@pjcollins  any idea why testes would be failing ?

mattleibow commented on src/Templates/src/templates/maui-mobile/MauiApp.1.csproj:
```diff
@@ -56,6 +56,8 @@
 	</ItemGroup>
 
 	<ItemGroup>
+		<PackageReference Include="Microsoft.Maui.Controls" Version="$(MauiVersion)" />
```
This is new. When updating package refs in the IDE or CLI, the implicit packages causes issues. So we are now explicit. We also have a new warning for existing net8 projects. In net9 we may only do explicit packages as implicit packages is not 100% recommended by anyone.

github-actions[bot] committed:
[create-pull-request] automated change (#14840)

Co-authored-by: github-actions[bot] <41898282+github-actions[bot]@users.noreply.github.com>

mattleibow committed:
Merge remote-tracking branch 'origin/net8.0' into merge-main-net8

mattleibow committed:
Merge remote-tracking branch 'origin/main' into merge-main-net8

mattleibow committed:
Update TemplateTests.cs

jonathanpeppers committed:
[android] MauiTextView doesn't need ViewAttachedToWindow (#14833)

Context: https://github.com/dotnet/maui/issues/12130
Context: https://github.com/angelru/CvSlowJittering

Profiling a customer sample, I noticed that ~2.4% of of the time was
spent subscribing to `ViewAttachedToWindow`:

    278.55ms (2.4%) mono.android!Android.Views.View.add_ViewAttachedToWindow(System.EventHandler`1<Android.Views.View/ViewAttachedToWindowEv

It appeared that for every `Label` on Android, we setup this event.
Additionally, Java would have to call into C# when the event fires:

    30.55ms (0.26%) mono.android!Android.Views.View.IOnAttachStateChangeListenerInvoker.n_OnViewAttachedToWindow_Landroid_view_View__mm_wra

However, we don't actually *do* anything in the event (added in
b6c3b533), so we should be able to just delete it?

    this.ViewAttachedToWindow += MauiTextView_ViewAttachedToWindow;
    //...
    private void MauiTextView_ViewAttachedToWindow(object? sender, ViewAttachedToWindowEventArgs e)
    {
    }

Doing this, dropped the call to ~0.02%:

    2.76ms (0.02%) mono.android!Android.Views.View.add_ViewAttachedToWindow(System.EventHandler`1<Android.Views.View/ViewAttachedToWindowEv

And the above `IOnAttachStateChangeListenerInvoker` call is gone
completely.

This should improve the performance of every `Label` on Android.

MSLukeWest committed:
Enabling tolerance adjustment in iOS color verification tests (#13945)

* Enabling tolerance adjustment in iOS color verification tests

* Resolving missed merge conflict

* Adding back changes removed by merge

---------

Co-authored-by: Luke Westendorf <lukewest@microsoft.com>

pjcollins committed:
[tests] Update timeout for iOS template runs (#14863)

Bumps the timeout for running iOS templates on simulator from 25 to 75
seconds.  We've been seeing intermittent failures in this test which are
likely caused by differences in simulator launch/startup times across
different test machines.

mattleibow committed:
Merge remote-tracking branch 'origin/main' into merge-main-net8
